src/Perception/src/ArucoDetector.cpp

```diff
@@ -296,3 +343,185 @@ void ArucoDetector::Impl::resetBuffers()
+                                             detectedMarkerTransVecs);
+
+    }
+    else
+    {
+        solvePnPMarkerPose(ids,
+                           corners,
+                           mLength,
+                           camMat,
+                           distCoeffs,
+                           ambiguityThreshold,
+                           detectedMarkerRotVecs,
+                           detectedMarkerTransVecs);
+    }
+}
+
+
+void ArucoDetector::Impl::solvePnPMarkerPose(std::vector<int>& ids,
+                                             std::vector<std::vector<cv::Point2f>>& corners,
+                                             const double& mLength,
+                                             const cv::Mat& camMat,
+                                             const cv::Mat& distCoeffs,
+                                             const double& ambiguityThreshold,
+                                             std::vector<cv::Vec3d>& detectedMarkerRotVecs,
+                                             std::vector<cv::Vec3d>& detectedMarkerTransVecs)
+{
+    constexpr auto printPrefix = "[ArucoDetector::Impl::solvePnPMarkerPose]";
+    if (mLength < 0)
+    {
+        log()->error("{} Invalid marker length.", printPrefix);
+    }
+
+    auto nrMarkers{ids.size()};
+
+    detectedMarkerRotVecs.resize(nrMarkers);
+    detectedMarkerTransVecs.resize(nrMarkers);
+
+    std::vector<int> unambiguousIDs;
+    for (auto mIdx = 0; mIdx < nrMarkers; mIdx++)
+    {
+        cv::Vec3d rVec, tVec;
+        bool solved = solvePnPSingleMarkerPose(corners[mIdx],
+                                               mLength,
+                                               camMat,
+                                               distCoeffs,
+                                               ambiguityThreshold,
+                                               rVec,
+                                               tVec);
+
+        if (solved)
+        {
+            unambiguousIDs.emplace_back(mIdx);
+        }
+
+        detectedMarkerRotVecs[mIdx] = rVec;
+        detectedMarkerTransVecs[mIdx] = tVec;
+    }
+
+    // if the two solutions of estimated rotations
+    // is too close and ambiguous, then mark the
+    // the marker as not detected
+    std::vector<int> newIds;
+    std::vector<std::vector<cv::Point2f>> newCorners;
+    std::vector<cv::Vec3d> newDetectedMarkerRotVecs;
+    std::vector<cv::Vec3d> newDetectedMarkerTransVecs;
+
+    for (const auto& mIdx : unambiguousIDs)
+    {
+        newIds.emplace_back(ids[mIdx]);
+        newCorners.emplace_back(corners[mIdx]);
+        newDetectedMarkerRotVecs.emplace_back(detectedMarkerRotVecs[mIdx]);
+        newDetectedMarkerTransVecs.emplace_back(detectedMarkerTransVecs[mIdx]);
+    }
+
+    ids = newIds;
+    corners = newCorners;
+    detectedMarkerRotVecs = newDetectedMarkerRotVecs;
+    detectedMarkerTransVecs = newDetectedMarkerTransVecs;
```

Comment:
Why not use directly the memory given by the user instead of dynamically allocating the memory twice?

Reply:
Suggested changes applied in https://github.com/ami-iit/bipedal-locomotion-framework/pull/503/commits/b1822e82a41d7c63b27f79ec7d81e16e721a0be9.